[docs] Document the new editing API (#4060)
mui/mui-x: docs/data/data-grid/editing-legacy/editing.md

---
title: Data Grid - Editing (legacy)
---

# Data Grid - Editing (legacy)

<p class="description">The data grid has built-in edit capabilities that you can customize to your needs.</p>

## Cell editing

> Cell editing allows editing the value of one cell at a time.

Reviewer: "Allow" is not generally recommended in this context. See: [Google developer style guide word list—enable](https://developers.google.com/style/word-list#enable). I have also added a note about this in [our style guide on Notion](https://www.notion.so/mui-org/Style-guide-2a957a4168a54d47b14bae026d06a24b) for future reference.

```suggestion
Cell editing lets you edit the value of one cell at a time.
```

All the same, this sentence feels redundant: "Cell editing lets you edit cells." I think it could be cut.

> Set the `editable` property in the `GridColDef` object to `true` to allow editing cells of this column.

Reviewer: Along with swapping out "allow," I would rearrange this sentence to place the most important clause in front:

```suggestion
To enable cell editing within a column, set the `editable` property in the `GridColDef` object to `true`.
```

- A <kbd class="key">Enter</kbd> keydown
- A <kbd class="key">Backspace</kbd> or <kbd class="key">Delete</kbd> keydown. It will also delete the value and stops the edit mode instantly.
- A keydown of any printable key, for instance `a`, `E`, `0`, or `$`
- A double click on the cell
- A call to `apiRef.current.setCellMode(id, field, 'edit')`.

  ```tsx
  /**
    * Set the cellMode of a cell.
    * @param GridRowId
    * @param string
    * @param 'edit' | 'view'
    */
  setCellMode: (id: GridRowId, field: string, mode: GridCellMode) => void;
  ```

### Stop editing

If a cell is in edit mode and has focus, any of the following interactions will stop the edit mode:

- A <kbd class="key">Escape</kbd> keydown. It will also roll back changes done in the value of the cell.
- A <kbd class="key">Tab</kbd> keydown. It will also save and goes to the next cell on the same row.
- A <kbd class="key">Enter</kbd> keydown. It will also save and goes to the next cell on the same column.
- A mousedown outside the cell
- A call to `apiRef.current.setCellMode(id, field, 'view')`.

### Control cell editability

> In addition to the `editable` flag on columns, control which cell is editable using the `isCellEditable` prop.

Reviewer: ```suggestion
In addition to the `editable` flag on columns, you can enable or disable editing of individual cells using the `isCellEditable` prop.
```

In this demo, only the rows with an even `Age` value are editable.
The editable cells have a green background for better visibility.

{{"demo": "IsCellEditableGrid.js", "bg": "inline"}}

### Controlled editing

The `editRowsModel` prop lets you control the editing state.

> You can handle the `onEditRowsModelChange` callback to control the `GridEditRowsModel` state.

Reviewer: I don't understand the usage of "handle" here. I think it would be simpler to just say "use":

```suggestion
You can use the `onEditRowsModelChange` callback to control the `GridEditRowsModel` state.
```

Author: In React there's a convention of naming props as `onXXX` if they are called when a particular, e.g. `xxx`, event is fired. The function passed to the prop will be called when the event is fired, so it "handles" the event. https://reactjs.org/docs/handling-events.html

{{"demo": "CellEditControlGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Saving nested structures

> If you are using a `valueGetter` to extract the value from a nested object, then a `valueSetter` also needs to be provided.

Reviewer: Swap passive for active voice.

```suggestion
If you are using a `valueGetter` to extract the value from a nested object, then you must also provide a `valueSetter`.
```

The first one receives the row object and must return the value to be displayed in the cell.
In the other side, the second one does the inverse, receiving the new value entered and returning the updated row.

The following demo shows how these two functions can be used:

{{"demo": "ValueGetterSetterGrid.js", "bg": "inline"}}

> Calling the `valueSetter` is the last step in the saving process.
> The [validation](/components/data-grid/editing/#client-side-validation) will still be called with the values before they pass through the setter.

### Client-side validation

> To validate the value in the cells, first add a `preProcessEditCellProps` callback to the [column definition](/api/data-grid/grid-col-def/) of the field to validate.

Reviewer: ```suggestion
To validate the value of a cell, add the `preProcessEditCellProps` callback to the [column definition](/api/data-grid/grid-col-def/) of the field you wish to validate.
```

> Once it is called, validate the value provided in `params.props.value`.

Reviewer: ```suggestion
Once it is called, you can validate the value provided in `params.props.value`.
```

> Then, return a new object contaning `params.props` and also the `error` attribute set to true or false.

Reviewer: ```suggestion
Then, return a new object containing `params.props` along with the `error` attribute set to true or false.
```

If the `error` attribute is true, the value will never be committed.

```tsx
const columns: GridColDef[] = [
  {
    field: 'firstName',
    preProcessEditCellProps: (params: GridPreProcessEditCellProps) => {
      const hasError = params.props.value.length < 3;
      return { ...params.props, error: hasError };
    },
  },
];
```

Here is an example implementing an email validation:

{{"demo": "ValidateRowModelControlGrid.js", "bg": "inline"}}

> Alternatively, you can use the `GridEditRowsModel` state mentioned in the [Controlled editing](#controlled-editing) section.
> However, one limitation of this approach is that it does not work with the `singleSelect` column type.

### Server-side validation

Server-side validation works like [client-side validation](#client-side-validation).

> The only difference is that when `preProcessEditCellProps` is called, a promise must be returned.

Reviewer: Active voice.

```suggestion
The only difference is that when you call `preProcessEditCellProps`, you must return a promise.
```

If you want to send the updated data to your server, you can use the `onCellEditCommit` which is fired just before committing the new cell value to the grid.

You can then decide if you want to send the whole row or only the modified fields.

{{"demo": "CellEditServerSidePersistence.js", "bg": "inline", "defaultCodeOpen": false}}

### Custom edit component

To customize the edit component of a column, use the `renderEditCell` attribute available in the `GridColDef`.

The demo lets you edit the ratings by double-clicking the cell.

{{"demo": "RenderRatingEditCellGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Edit using external button [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

You can override the default [start editing](#start-editing) triggers using the [`event.defaultMuiPrevented`](/components/data-grid/events#disabling-the-default-behavior) on the synthetic React events.

{{"demo": "StartEditButtonGrid.js", "bg": "inline", "disableAd": true}}

### Events [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

The editing feature leverages the event capability of the grid and the apiRef.

> The following events can be imported and used to customize the edition:

Reviewer: Active voice.

```suggestion
You can import the following events to customize the editing experience:
```

> - `cellEditStart`: emitted when the cell turns to edit mode.
> - `cellEditStop`: emitted when the cell turns back to view mode.
> - `cellEditCommit`: emitted when the new value is committed.
> - `editCellPropsChange`: emitted when the props passed to the edit cell component are changed.

Reviewer: This is an odd usage of the word "emitted," which generally connotes a sensation like a smell or a noise. 

I think "triggered" would be more accurate.

```suggestion
- `cellEditStart`: triggered when a cell enters edit mode.
- `cellEditStop`: triggered when a cell returns to view mode.
- `cellEditCommit`: triggered when a new value is committed.
- `editCellPropsChange`: triggered when the props passed to the edit cell component are changed.
```

Reviewer 2: I think it is derived from the term "Event emitter" which is very common.
So `emit and event` feels right to me.
But here we are talking about the DOM interaction which fires the event, so I'm fine with "triggered when"

Reviewer 2: Would work for me

> Catching events can be used to add a callback after an event while ignoring its triggers.

Reviewer: Active voice, and singular/plural alignment.

```suggestion
You can use a catching event to add a callback after an event while ignoring its triggers.
```

Author: I changed to "You can use event catching...". A "catching event" doesn't make much sense in this context.

> The demo shows how to catch the start & end edit events to log which cell is editing in an info message:

Reviewer: Let's spell out "and" instead of using an ampersand. Also, I think the verb conjugation we want in the latter half of the sentence is "has been edited," but I could be wrong if I've misunderstood the context. 

```suggestion
The demo shows how to catch the start and end edit events in order to log which cell has been edited in an info message:
```

Reviewer: Unless it's more accurate to use the plural, in which case we'd want to say "which cells have been edited."

Author: The demo only logs the current cell so the singular version is more correct.

{{"demo": "CatchEditingEventsGrid.js", "bg": "inline", "disableAd": true}}

## Row editing

> Row editing allows to edit all the cells of a row at once.

Reviewer: ```suggestion
Row editing lets you edit all cells in a given row simultaneously.
```

> It is based on the [cell editing](/components/data-grid/editing/#cell-editing), thus most of the features are also supported.

Reviewer: ```suggestion
It supports most of the same features as those available for [cell editing](/components/data-grid/editing/#cell-editing).
```

Return a new object contaning `params.props` and the `error` attribute with the validation status.
Once at the least one field has the `error` attribute equals to true no new value will be committed.

**Note:** For server-side validation, the same [approach](#server-side-validation) from the cell editing can be used. The data grid will wait for all promises to resolve before commiting.

The following demo requires a value for the "Payment method" column if the "Is paid?" column was checked.

{{"demo": "ConditionalValidationGrid.js", "disableAd": true, "bg": "inline", "defaultCodeOpen": false}}

> The conditional validation can also be implemented with the [controlled editing](#controlled-editing-2).
> This approach can be used in the free version of the DataGrid.
> The only limitation is that it does not work with the `singleSelect` column type.

### Control with external buttons [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

You can [disable the default behavior](/components/data-grid/events/#disabling-the-default-behavior) of the grid and control the row edit using external buttons.

> Here is shown how a full-featured CRUD can be created.

Author: Updated. I only removed "app" because it's not a real app. It's more of a component that lets create, read, update, etc. records like in a business application.

{{"demo": "FullFeaturedCrudGrid.js", "bg": "inline", "disableAd": true}}

### Saving rows with nested structures

> Saving columns that make use of `valueGetter` can be done adding a `valueSetter`.

Reviewer: ```suggestion
You can save columns that make use of `valueGetter` by adding a `valueSetter`.
```

> The same [approach](/components/data-grid/editing/#saving-nested-structures) from the cell editing mode can be used here.

Reviewer: Add the relevant info here.

```suggestion
The 'valueGetter` receives the row object and must return the value to be displayed in the cell.
The 'valueSetter`, in turn, receives the new value and returns the updated row.
```

Author: It's the same thing from the cell editing. I don't want to duplicate content.

Reviewer: From the perspective of the reader's experience, it can be confusing to have to jump around within the document to find the relevant info. It would be easy to lose track of whether you were in the "cell" sections or the "row" sections to begin with if you have jump between them.

Note that the `valueSetter` will be called for each field.

### Server-side persistence [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

> If you want to send the updated data to your server, you can use the `onRowEditCommit` which is fired just before committing the new cell value to the grid.

Reviewer: ```suggestion
If you want to send the updated data to your server, you can use `onRowEditCommit` which is fired just before committing the new cell value to the grid.
```

> To access the new values for the row, use `apiRef.current.getEditRowsModel` to get all rows in edit mode, then use the id provided to get only the values for the row that was committed.

Reviewer: ```suggestion
To access the new values for the row, use `apiRef.current.getEditRowsModel` to enable edit mode on all rows, then use the ID provided to get only the values for the row that was committed.
```

> You can then decide if you want to send the whole row or only the modified fields, by checking them against the previous row values.

Reviewer: ```suggestion
You can then decide if you want to send the whole row or only the modified fields by checking them against the previous row values.
```

{{"demo": "RowEditServerSidePersistence.js", "disableAd": true, "bg": "inline", "defaultCodeOpen": false}}

### Events [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

> The following events can be imported and used to customize the row edition:

Reviewer: ```suggestion
You can import the following events to customize the editing experience:
```

- `rowEditStart`: emitted when the row turns to edit mode.
- `rowEditStop`: emitted when the row turns back to view mode.
- `rowEditCommit`: emitted when the new row values are committed.

> - `editCellPropsChange`: emitted when the props passed to an edit cell component are changed.

Reviewer: - `rowEditStart`: triggered when a row enters edit mode.
- `rowEditStop`: triggered when a row returns to view mode.
- `rowEditCommit`: triggered when new row values are committed.
- `editCellPropsChange`: triggered when the props passed to the edit cell component are changed.

## Cell editing

To enable cell editing within a column, set the `editable` property in the `GridColDef` object to `true` to allow editing cells of this column.

```tsx
<DataGrid columns={[{ field: 'name', editable: true }]} />
```

{{"demo": "BasicEditingGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Start editing

If a cell is editable and has focus, any of the following interactions will start the edit mode:

- A <kbd class="key">Enter</kbd> keydown
- A <kbd class="key">Backspace</kbd> or <kbd class="key">Delete</kbd> keydown. It will also delete the value and stops the edit mode instantly.
- A keydown of any printable key, for instance `a`, `E`, `0`, or `$`
- A double click on the cell
- A call to `apiRef.current.setCellMode(id, field, 'edit')`.

  ```tsx
  /**
    * Set the cellMode of a cell.
    * @param GridRowId
    * @param string
    * @param 'edit' | 'view'
    */
  setCellMode: (id: GridRowId, field: string, mode: GridCellMode) => void;
  ```

### Stop editing

If a cell is in edit mode and has focus, any of the following interactions will stop the edit mode:

> - A <kbd class="key">Escape</kbd> keydown. It will also roll back changes done in the value of the cell.
> - A <kbd class="key">Tab</kbd> keydown. It will also save and goes to the next cell on the same row.
> - A <kbd class="key">Enter</kbd> keydown. It will also save and goes to the next cell on the same column.
> - A mousedown outside the cell
> - A call to `apiRef.current.setCellMode(id, field, 'view')`.

Reviewer: Same as above—actions over objects, punctuation consistency, present tense.

```suggestion
- Pressing <kbd class="key">Escape</kbd>—this also rolls back any changes made
- Pressing <kbd class="key">Tab</kbd>—this also saves any changes made and moves the focus to the next cell in the row
- Pressing <kbd class="key">Enter</kbd>—this also saves any changes made and moves the focus to the next cell in the column
- Clicking the mouse outside of the cell
- Making a call to `apiRef.current.setCellMode(id, field, 'view')`
```

### Control cell editability

In addition to the `editable` flag on columns, you can enable or disable editing of individual cells using the `isCellEditable` prop.

In this demo, only the rows with an even `Age` value are editable.
The editable cells have a green background for better visibility.

{{"demo": "IsCellEditableGrid.js", "bg": "inline"}}

> ### Controlled editing

Reviewer: Clarity and style consistency.

```suggestion
### Control cell editing state
```

The `editRowsModel` prop lets you control the editing state.

> You can use the `onEditRowsModelChange` callback to control the `GridEditRowsModel` state.

Reviewer: Add colon to indicate that this paragraph offers information about the code sample that follows.

```suggestion
You can use the `onEditRowsModelChange` callback to control the `GridEditRowsModel` state:
```

{{"demo": "CellEditControlGrid.js", "bg": "inline", "defaultCodeOpen": false}}

> ### Saving nested structures

Reviewer: Verb tweak for style consistency across the document.

```suggestion
### Save nested structures
```

> To avoid this inconsistency, enable the `preventCommitWhileValidating` flag available in the experimental features.
>
> ```jsx
> <DataGrid experimentalFeatures={{ preventCommitWhileValidating: true }} />
> ```
>
> It's labeled as experimental because in v6 this flag will become the default behavior.

This demo shows how you can validate a username asynchronously and prevent the user from committing the value while validating.
It's using `DataGridPro` but the same approach can be used with `DataGrid`.

{{"demo": "ValidateServerNameGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Server-side persistence

If you want to send the updated data to your server, you can use the `onCellEditCommit` which is fired just before committing the new cell value to the grid.

You can then decide if you want to send the whole row or only the modified fields.

{{"demo": "CellEditServerSidePersistence.js", "bg": "inline", "defaultCodeOpen": false}}

### Custom edit component

To customize the edit component of a column, use the `renderEditCell` attribute available in the `GridColDef`.

> The demo lets you edit the ratings by double-clicking the cell.

Reviewer: Clarity and colon.

```suggestion
This demo shows how you can edit ratings by double-clicking on the cell:
```

- `cellEditStop`: triggered when a cell returns to view mode.
- `cellEditCommit`: triggered when a new value is committed.
- `editCellPropsChange`: triggered when a props passed to the edit cell component are changed.

You can use event catching to add a callback after an event while ignoring its triggers.

The demo shows how to catch the start and end edit events in order to log which cell has been edited in an info message:

{{"demo": "CatchEditingEventsGrid.js", "bg": "inline", "disableAd": true}}

## Row editing

Row editing lets you edit all cells in a given row simultaneously.
It supports most of the same features as those available for [cell editing](/components/data-grid/editing/#cell-editing).
To enable it, change the edit mode to `"row"` using the `editMode` prop, then set to `true` the `editable` property in the `GridColDef` object of those columns that should be editable.

```tsx
<DataGrid editMode="row" columns={[{ field: 'name', editable: true }]} />
```

{{"demo": "BasicRowEditingGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Start editing

You can start editing a cell using any of the following interactions:

> - A <kbd class="key">Enter</kbd> keydown
> - A double click on the cell
> - A call to `apiRef.current.setRowMode(id, 'edit')`.

Reviewer: Actions over objects. Remove period.

```suggestion
- Pressing <kbd class="key">Enter</kbd>
- Double-clicking on the cell
- Making a call to `apiRef.current.setRowMode(id, 'edit')`
```

  ```tsx
  /**
    * Sets the mode of a row.
    * @param {GridRowId} id The id of the row.
    * @param {GridRowMode} mode Can be: `"edit"`, `"view"`.
    */
  setRowMode: (id: GridRowId, mode: GridRowMode) => void;
  ```

### Stop editing

You can stop editing a cell using any of the following interactions:

> - A <kbd class="key">Escape</kbd> keydown. This will also roll back any changes made to the row.
> - A <kbd class="key">Enter</kbd> keydown. This will also save the value and move the focus to the cell in the next row of the same column.
> - A mouse click outside the row
> - A call to `apiRef.current.setRowMode(id, 'view')`.

Reviewer: Actions over objects; style consistency; present tense.

```suggestion
- Pressing <kbd class="key">Escape</kbd>—this also rolls back any changes made to the row
- Pressing <kbd class="key">Enter</kbd>—this also saves any changes made and moves the focus to the next cell same column
- Clicking the mouse outside of the row
- Making a call to `apiRef.current.setRowMode(id, 'view')`
```

> ### Controlled editing

Reviewer: Clarity and style consistency.

```suggestion
### Control row editing state
```

The `editRowsModel` prop lets you control the editing state.

> You can handle the `onEditRowsModelChange` callback to control the `GridEditRowsModel` state.

Reviewer: ```suggestion
You can handle the `onEditRowsModelChange` callback to control the `GridEditRowsModel` state:
```

{{"demo": "RowEditControlGrid.js", "bg": "inline", "defaultCodeOpen": false}}

### Conditional validation [<span class="plan-pro"></span>](https://mui.com/store/items/material-ui-pro/)

> Having all cells of a row in edit mode allows validating a field based on the value of another one

Reviewer: Replace this line with 215 as they say the same thing, but 215 is more clear.

```suggestion
When all cells in a row are in edit mode, you can validate fields by comparing their values.
```

To do this, you will first need to add a `preProcessEditCellProps` callback to the [column definition](/api/data-grid/grid-col-def/).

> When all cells in a row are in edit mode, you can validate fields by comparing their values.

Reviewer: Move this line to the beginning of the paragraph (213).

```suggestion
```

Return a new object contaning `params.props` and the `error` attribute with the validation status.

> Once at the least one field has the `error` attribute equals to true no new value will be committed.

Reviewer: Grammar. Also, I believe `true` should get back ticks here.

```suggestion
Once at the least one field has the `error` attribute set to `true`, no new value will be committed.
```

> ⚠ Server-side validation works like [client-side validation](#client-side-validation).
> The only difference is that when you call `preProcessEditCellProps`, you must return a promise.
> Once the value is validated in the server, the promise should be resolved with a new object containing the `error` attribute set to true or false.
> The grid will wait for the promise to be resolved before exiting the edit mode.

> The following demo requires a value for the **Payment method** column if the **Is paid?** column is checked.

Reviewer: ```suggestion
The following demo requires a value for the **Payment method** column if the **Is paid?** column is checked:
```
